src/Nethermind/Nethermind.Evm/Tracing/GethStyle/GethLikeBlockFileTracer.cs

```diff
@@ -0,0 +1,120 @@
+using System.Linq;
+using System.Text.Json;
+using Nethermind.Core.Crypto;
+using Nethermind.Core;
+using Nethermind.Core.Extensions;
+using System.IO.Abstractions;
+
+namespace Nethermind.Evm.Tracing.GethStyle;
+
+public class GethLikeBlockFileTracer : BlockTracerBase<GethLikeTxTrace, GethLikeTxFileTracer>
+{
+    private const string _alphabet = "abcdefghijklmnopqrstuvwxyz0123456789";
+
+    private readonly Block _block;
+    private Stream _file;
+    private readonly string _fileNameFormat;
+    private readonly List<string> _fileNames = new();
+    private IFileSystem _fileSystem;
+    private Utf8JsonWriter _jsonWriter;
+    private readonly GethTraceOptions _options;
+    private readonly JsonSerializerOptions _serializerOptions = new();
+
+    public GethLikeBlockFileTracer(Block block, GethTraceOptions options, IFileSystem fileSystem) : base(options?.TxHash)
+    {
+        _block = block ?? throw new ArgumentNullException(nameof(block));
+        _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+        _options = options ?? throw new ArgumentNullException(nameof(options));
+
+        var hash = _block.Hash.Bytes[..4].ToHexString(true);
+
+        _fileNameFormat = _fileSystem.Path.Combine(_fileSystem.Path.GetTempPath(), $"block_{hash}-{{0}}-{{1}}-{{2}}.jsonl");
+
+        _serializerOptions.Converters.Add(new GethLikeTxTraceJsonConverter());
+    }
+
+    public IReadOnlyCollection<string> FileNames => _fileNames.AsReadOnly();
+
+    public override void EndBlockTrace()
+    {
+        base.EndBlockTrace();
+
+        DisposeFileStreamIfAny();
+    }
+
+    protected override void AddTrace(GethLikeTxTrace trace) { }
+
+    protected override GethLikeTxTrace OnEnd(GethLikeTxFileTracer txTracer)
+    {
+        var trace = txTracer.BuildResult();
+
+        JsonSerializer.Serialize(_jsonWriter,
+            new
+            {
+                output = trace.ReturnValue.ToHexString(true),
+                gasUsed = $"0x{trace.Gas:x}"
+            },
+            _serializerOptions);
+
+        DisposeFileStreamIfAny();
+
+        return trace;
+    }
+
+    protected override GethLikeTxFileTracer OnStart(Transaction? tx)
+    {
+        // Ensure the current file stream is disposed in case of API misuse
+        DisposeFileStreamIfAny();
+
+        _fileNames.Add(GetFileName(tx.Hash));
+
+        _file = _fileSystem.File.OpenWrite(_fileNames.Last());
+        _jsonWriter = new Utf8JsonWriter(_file);
+
+        return new(DumpTraceEntry, _options);
+    }
+
+    private void DisposeFileStreamIfAny()
+    {
+        _jsonWriter?.Dispose();
+        _file?.Dispose();
+
+        _file = null;
+        _jsonWriter = null;
+    }
+
+    private void DumpTraceEntry(GethTxFileTraceEntry entry)
+    {
+        JsonSerializer.Serialize(_jsonWriter, entry, _serializerOptions);
+
+        _jsonWriter.Flush();
+        _jsonWriter.Reset();
+        _jsonWriter.WriteRawValue(Environment.NewLine, true);
```

Comment:
Why new line? Maybe make it optional? If we stream to a network socket we don't need human readable format.

Reply:
What's the point of taking this to an extension method? It's only used here in this particular case.

Reply:
To name it so we know the intention. Can be a normal method.

Reply:
I put some comments instead.